// ts-graphql-plugin/lib/typegen-addons/testing/addon-tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAddonTester = void 0;
const graphql_1 = require("graphql");
const testing_extractor_1 = require("../../analyzer/testing/testing-extractor");
const type_generator_1 = require("../../analyzer/type-generator");
function createTestingTypeGenerator({ files = [], tag = '', addonFactories = [], }) {
    const extractor = (0, testing_extractor_1.createTesintExtractor)(files, true);
    const generator = new type_generator_1.TypeGenerator({
        prjRootPath: '',
        tag,
        addonFactories,
        extractor,
        debug: () => { },
    });
    return generator;
}
class AddonTester {
    constructor(facory, options) {
        this.facory = facory;
        this.options = options;
    }
    generateTypes({ files, schemaSDL }) {
        const schema = (0, graphql_1.buildSchema)(schemaSDL);
        const generator = createTestingTypeGenerator({ files, addonFactories: [this.facory], tag: this.options.tag });
        const { errors, outputSourceFiles } = generator.generateTypes({ files: files.map(f => f.fileName), schema });
        return { errors, outputSourceFiles };
    }
}
function createAddonTester(factory, options = {}) {
    return new AddonTester(factory, options);
}
exports.createAddonTester = createAddonTester;
