// ts-graphql-plugin/lib/register-hooks/register-typescript.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerTypeScript = void 0;
const fs_1 = __importDefault(require("fs"));
function registerTypeScript() {
    let defaultCompileOptions;
    require.extensions['.ts'] = (module, fileName) => {
        const ts = require('typescript');
        if (!defaultCompileOptions) {
            defaultCompileOptions = ts.getDefaultCompilerOptions();
        }
        const content = fs_1.default.readFileSync(fileName, 'utf8');
        const { outputText } = ts.transpileModule(content, {
            fileName,
            compilerOptions: {
                ...defaultCompileOptions,
                noEmit: true,
                esModuleInterop: true,
                target: ts.ScriptTarget.ES2019,
                module: ts.ModuleKind.CommonJS,
            },
            reportDiagnostics: false,
        });
        module._compile(outputText, fileName);
    };
}
exports.registerTypeScript = registerTypeScript;
