// ts-graphql-plugin/lib/schema-manager/file-schema-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSchemaManager = void 0;
const path_1 = __importDefault(require("path"));
const graphql_1 = require("graphql");
const schema_manager_1 = require("./schema-manager");
function extractIntrospectionContentFromJson(jsonObject) {
    if (jsonObject.data) {
        return jsonObject.data;
    }
    else {
        return jsonObject;
    }
}
class FileSchemaManager extends schema_manager_1.SchemaManager {
    constructor(_host, options) {
        super(_host);
        this._host = _host;
        this._schemaPath = options.path;
    }
    getBaseSchema() {
        if (!this._schemaPath || typeof this._schemaPath !== 'string')
            return null;
        try {
            const resolvedSchmaPath = this.getAbsoluteSchemaPath(this._host.getProjectRootPath(), this._schemaPath);
            this.log('Read schema from ' + resolvedSchmaPath);
            const isExists = this._host.fileExists(resolvedSchmaPath);
            if (!isExists)
                return null;
            if (this._schemaPath.endsWith('.graphql') || this._schemaPath.endsWith('.gql')) {
                const sdl = this._host.readFile(resolvedSchmaPath, 'utf-8');
                return sdl ? (0, graphql_1.buildSchema)(sdl) : null;
            }
            else {
                const introspectionContents = this._host.readFile(resolvedSchmaPath, 'utf-8');
                return introspectionContents
                    ? (0, graphql_1.buildClientSchema)(extractIntrospectionContentFromJson(JSON.parse(introspectionContents)))
                    : null;
            }
        }
        catch (err) {
            this.log('Fail to read schema file...');
            this.log(err instanceof Error ? err.message : `Unknown error: ${err}`);
            return null;
        }
    }
    async waitBaseSchema() {
        return this.getBaseSchema();
    }
    getAbsoluteSchemaPath(projectRootPath, schemaPath) {
        if (path_1.default.isAbsolute(schemaPath))
            return schemaPath;
        return path_1.default.resolve(projectRootPath, schemaPath);
    }
    startWatch(interval = 100) {
        const resolvedSchmaPath = this.getAbsoluteSchemaPath(this._host.getProjectRootPath(), this._schemaPath);
        this._watcher = this._host.watchFile(resolvedSchmaPath, () => {
            this.log('Change schema file.');
            this.emitChange();
        }, interval);
    }
    closeWatch() {
        if (this._watcher)
            this._watcher.close();
    }
}
exports.FileSchemaManager = FileSchemaManager;
