// ts-graphql-plugin/lib/graphql-language-service-adapter/testing/adapter-fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdapterFixture = void 0;
const ts_ast_util_1 = require("../../ts-ast-util");
const graphql_language_service_adapter_1 = require("../graphql-language-service-adapter");
const testing_language_service_1 = require("../../ts-ast-util/testing/testing-language-service");
class AdapterFixture {
    constructor(sourceFileName, schema) {
        const { languageService, languageServiceHost } = (0, testing_language_service_1.createTestingLanguageServiceAndHost)({
            files: [{ fileName: sourceFileName, content: '' }],
        });
        this._sourceFileName = sourceFileName;
        this._langServiceHost = languageServiceHost;
        this.langService = languageService;
        this.adapter = new graphql_language_service_adapter_1.GraphQLLanguageServiceAdapter((0, ts_ast_util_1.createScriptSourceHelper)({ languageService, languageServiceHost }), {
            schema: schema || null,
            removeDuplicatedFragments: true,
        });
    }
    get source() {
        return this._langServiceHost.getFile(this._sourceFileName).content;
    }
    set source(content) {
        this._langServiceHost.updateFile(this._sourceFileName, content);
    }
}
exports.AdapterFixture = AdapterFixture;
